--- asf-strategy.hpp	2017-09-29 23:51:21.578926000 -0400
+++ new/asf-strategy.hpp	2017-09-29 23:52:42.858868000 -0400
@@ -76,6 +76,39 @@
 
   void
   sendNoRouteNack(const Face& inFace, const Interest& interest, const shared_ptr<pit::Entry>& pitEntry);
+  
+  virtual void 
+  found(const Face& inFace, const shared_ptr<pit::Entry>& pitEntry, const Interest& interest, const Data& data);
+
+  virtual void 
+  notfound(const Face& inFace, const shared_ptr<pit::Entry>& pitEntry, const Interest& interest);
+  
+  virtual void 
+  Trace(const Face& inFace, const Interest& interest, const shared_ptr<pit::Entry>& pitEntry);
+  
+  virtual void
+  Cache_check(const Face& inFace, const shared_ptr<pit::Entry>& pitEntry, const ndn::Name name);
+  
+  virtual void
+  multi_process(const Face& inFace, const Interest& interest, const shared_ptr<pit::Entry>& pitEntry);
+  
+  virtual void
+  single_process(const Face& inFace, const Interest& interest, const shared_ptr<pit::Entry>& pitEntry);
+  
+  virtual void
+  S_ForwardIt(const Face& inFace, const Interest& interest, const shared_ptr<pit::Entry>& pitEntry, const ndn::Name name);
+  
+  virtual void
+  M_ForwardIt(const Face& inFace, const Interest& interest, const shared_ptr<pit::Entry>& pitEntry, const ndn::Name name);
+  
+  const ndn::Name GetLookupName(const Interest& interest);
+  
+  private:	
+	Forwarder& m_forwarder;
+	bool in = true;
+	ndn::Name nx = "name";
+	int i = 0;
+	int j = 0;
 
 public:
   static const Name STRATEGY_NAME;
