--- nack-header.cpp	2017-09-29 23:34:19.326898528 -0400
+++ new.cpp	2017-03-06 12:52:47.426902000 -0500
@@ -39,6 +39,12 @@
   case NackReason::NO_ROUTE:
     os << "NoRoute";
     break;
+  case NackReason::CACHE_LOCAL:
+    os << "LocalCache";
+    break;
+  case NackReason::PRODUCER_LOCAL:
+    os << "LocalProducer";
+    break;
   default:
     os << "None";
     break;
@@ -122,6 +128,8 @@
   case NackReason::CONGESTION:
   case NackReason::DUPLICATE:
   case NackReason::NO_ROUTE:
+  case NackReason::PRODUCER_LOCAL:
+  case NackReason::CACHE_LOCAL:
     return m_reason;
   default:
     return NackReason::NONE;
